ToolKit/doc/man/nroff/rename.2
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)rename.2	8.1 (Berkeley) 6/4/93
.\"	$Id: rename.2,v 1.1.1.1 2006/05/30 06:12:12 hhzhou Exp $
.\"
.Dd June 4, 1993
.Dt RENAME 2
.Os BSD 4.2
.Sh NAME
.Nm rename
.Nd change the name of a file
.Sh SYNOPSIS
.Fd #include <stdio.h>
.Ft int
.Fn rename "const char *from" "const char *to"
.Sh DESCRIPTION
.Fn Rename
causes the link named
.Fa from
to be renamed as
.Fa to .
If 
.Fa to
exists, it is first removed.
Both 
.Fa from
and
.Fa to
must be of the same type (that is, both directories or both
non-directories), and must reside on the same file system.
.Pp
.Fn Rename
guarantees that an instance of
.Fa to
will always exist, even if the system should crash in
the middle of the operation.
.Pp
If the final component of
.Fa from
is a symbolic link,
the symbolic link is renamed,
not the file or directory to which it points.
.\".Sh CAVEAT
.\"The system can deadlock if a loop in the file system graph is present.
.\"This loop takes the form of an entry in directory
.\".Ql Pa a ,
.\"say
.\".Ql Pa a/foo ,
.\"being a hard link to directory
.\".Ql Pa b ,
.\"and an entry in
.\"directory
.\".Ql Pa b ,
.\"say
.\".Ql Pa b/bar ,
.\"being a hard link
.\"to directory
.\".Ql Pa a .
.\"When such a loop exists and two separate processes attempt to
.\"perform
.\".Ql rename a/foo b/bar
.\"and
.\".Ql rename b/bar a/foo ,
.\"respectively, 
.\"the system may deadlock attempting to lock
.\"both directories for modification.
.\"Hard links to directories should be
.\"replaced by symbolic links by the system administrator.
.Sh RETURN VALUES
A 0 value is returned if the operation succeeds, otherwise
.Fn rename
returns -1 and the global variable 
.Va errno
indicates the reason for the failure.
.Sh ERRORS
.Fn Rename
will fail and neither of the argument files will be
affected if:
.Bl -tag -width ENAMETOOLONG
.It Bq Er ENAMETOOLONG
A component of either pathname exceeded 255 characters,
or the entire length of either path name exceeded 1023 characters.
.It Bq Er ENOENT
A component of the
.Fa from
path does not exist,
or a path prefix of
.Fa to
does not exist.
.It Bq Er EACCES
A component of either path prefix denies search permission.
.It Bq Er EACCES
The requested link requires writing in a directory with a mode
that denies write permission.
.It Bq Er EPERM
The directory containing
.Fa from
is marked sticky,
and neither the containing directory nor
.Fa from
are owned by the effective user ID.
.It Bq Er EPERM
The
.Fa to
file exists,
the directory containing
.Fa to
is marked sticky,
and neither the containing directory nor
.Fa to
are owned by the effective user ID.
.It Bq Er ELOOP
Too many symbolic links were encountered in translating either pathname.
.It Bq Er ENOTDIR
A component of either path prefix is not a directory.
.It Bq Er ENOTDIR
.Fa from
is a directory, but
.Fa to
is not a directory.
.It Bq Er EISDIR
.Fa to
is a directory, but
.Fa from
is not a directory.
.It Bq Er EXDEV
The link named by
.Fa to
and the file named by
.Fa from
are on different logical devices (file systems).  Note that this error
code will not be returned if the implementation permits cross-device
links.
.It Bq Er ENOSPC
The directory in which the entry for the new name is being placed
cannot be extended because there is no space left on the file
system containing the directory.
.It Bq Er EDQUOT
The directory in which the entry for the new name
is being placed cannot be extended because the
user's quota of disk blocks on the file system
containing the directory has been exhausted.
.It Bq Er EIO
An I/O error occurred while making or updating a directory entry.
.It Bq Er EROFS
The requested link requires writing in a directory on a read-only file
system.
.It Bq Er EFAULT
.Em Path
points outside the process's allocated address space.
.It Bq Er EINVAL
.Fa From
is a parent directory of
.Fa to ,
or an attempt is made to rename
.Ql \&.
or
.Ql \&.. .
.It Bq Er ENOTEMPTY
.Fa To
is a directory and is not empty.
.El
.Sh SEE ALSO
.Xr open 2 ,
.Xr symlink 7
.Sh STANDARDS
The
.Fn rename
function call is expected to conform to 
.St -p1003.1-90 .


ToolKit/lib/libc/sys/rename.c
/*
 * Copyright (c) 1999, 2000
 * Intel Corporation.
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * 3. All advertising materials mentioning features or use of this software must
 *    display the following acknowledgement:
 * 
 *    This product includes software developed by Intel Corporation and its
 *    contributors.
 * 
 * 4. Neither the name of Intel Corporation or its contributors may be used to
 *    endorse or promote products derived from this software without specific
 *    prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY INTEL CORPORATION AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL INTEL CORPORATION OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 */

/*++

Module Name:

    rename.c
    
Abstract:

    Functions implementing the standard "rename" system call interface


Revision History

--*/


#include <efi_interface.h>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/param.h>
#include <sys/types.h>
#define KERNEL
#include <errno.h>
#undef KERNEL
#include "./filedesc.h"

#include <libc_debug.h>
#include <assert.h>

//
//  Name:
//      rename
//
//  Description:
//      BSD close interface for libc
//
//  Arguments:
//      From: path of existing name  To: path of new name
//
//  Returns:
//      0 on success, -1 otherwise
//

int
rename (const char *from, const char *to)
{
    EFI_STATUS      Status       = EFI_SUCCESS;
    int             ret          = 0;
	char			*s;
	int				retErrno	 = 0;
	int				fd			 = -1;
    wchar_t         *wcspath  	 = NULL;
    VOID            *DevSpecific = NULL;
	struct 			stat st;

	/*
	 *	Sanity check args 
	 */

	if (( from == NULL ) || (to == NULL) || (strlen (from) == 0) || (strlen (to) == 0) )
	{
   		DPRINT((L"rename: failed - null or zero length path \n"));
		retErrno = EINVAL;
		ret = -1;
		goto Done;
	}

	if ( (strlen (from) > MAXPATHLEN) || (strlen (to) > MAXPATHLEN) )
	{
   		DPRINT((L"rename: failed - path too long \n"));
		retErrno = ENAMETOOLONG;
		ret = -1;
		goto Done;
	}

	/*
	 *	Can't rename special dirs
	 */

	if ( (strcmp (to, ".") == 0) || (strcmp (to, "..") == 0) || (strcmp (from, ".") == 0)	||
		(strcmp (from, "..") == 0) )	
	{
   		DPRINT((L"rename: failed - '.' and '..' illegal \n"));
		retErrno = EINVAL;
		ret = -1;
		goto Done;
	}


	/*
	 *	EFI does not allow 'to' to have a device in it's path
	 */

    for (s = (char *)to; *s; s += 1) 
    {
        if (*s == ':') 
        {
	   		DPRINT((L"rename: device not allowed in new name \n"));
			retErrno = EXDEV;
			ret = -1;
            goto Done;
        }
    }

	/*
	 *	Fail if 'from' does not exist or 'to' already exists
	 */

	fd = open (from, O_RDWR);
	if (fd == -1)
	{
   		DPRINT((L"rename: error opening source \n"));
		retErrno = ENOENT;
		ret = -1;
		goto Done;
	}

	if (stat (to, &st) == 0)
	{
   		DPRINT((L"rename: new name already exists \n"));
		retErrno = EACCES;
		ret = -1;
		goto Done;
	}

	if ( _LIBC_GetOpenFileDevSpecific( fd, &DevSpecific ) != 0 ) {
		retErrno = ENOENT;
		ret = -1;
		goto Done;
	}

	wcspath = _ConvertToEfiPathname( to );
	if ( !wcspath ) {
		retErrno = ENOMEM;
		goto Done;
	}


    Status = EFI_RenameFile( DevSpecific, wcspath );
	if ( EFI_ERROR( Status ) ) 
	{
   		DPRINT((L"rename: EFI_RenameFile returned 0x%x\n",Status));
		retErrno = _EfiErrToErrno( Status );
		ret = -1;
		/* fall through and clean up */
	}

Done:

	if (fd != -1) close ( fd );
	if (wcspath) free( wcspath );

	errno=retErrno;

    return ret;
}

